### PR TITLE
Add packages: mosh

### PR DESCRIPTION
**Pull request by healther**

### CODE REVIEW

File: var/spack/repos/builtin/packages/mosh/package.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#
+# This is a template package file for Spack.  We've put "FIXME"
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mosh/package.py]: remve that

**Comment by citibeth** [comment]
Looks like a neat program.  Have you tried it?  Should I try it?  Does MacPorts have it, or do I need to start Mac-Spacking?  Do ^A, etc. work (doesn't work in GNU Screen...)

**Comment by healther** [comment]
I can only point you to @muffgaga , I only made the PRs from our local repository, sorry.

**Comment by hartzell** [comment]
It *is* neat.  It has a very specific problem set in mind: unreliable network connections.  Their examples include working via wifi on a commuter train and/or wandering around a campus, they're looking for a way to avoid dropping connections.  I thought it would be a great way for me to keep all of my work alive as I moved between builds and etc....

Part of its solution to the 'crappy connection' problem involves maintaining screen state on both sides of the link and only sending updates.  This makes [scrollback problematic](https://github.com/mobile-shell/mosh/issues/122), enough so that I gave up and went back to using tmux when I thought far enough ahead.

It installs easily on macs, using any of the package managers, info about halfway down here:  https://mosh.org/.

I should probably poke at it again and see how it's grown....

ps. control-a *does* work in screen, it just doesn't do what you (or I 😄...) want it to.  I've been using emacs too long to get over it and for some reason "losing" control-b to tmux doesn't bother me as much....

**Comment by healther** [comment]
I'm not sure why the code coverage went down, if you want further changes here, please feel free to tell me.

**Comment by citibeth** [comment]
> I'm not sure why the code coverage went down,

Because Spack got longer with a new package being added.

**Comment by adamjstewart** [comment]
@alalazo I didn't think the unit tests cared about builtin packages?

**Comment by alalazo** [comment]
@adamjstewart It doesn't, but for some reason codecov detects a change in coverage (-1 line) in `modules.yaml`. Can you try rebase this on the latest tip of develop?

**Comment by alalazo** [comment]
@citibeth We only monitor the framework with unit tests. Tests of packages are not yet included :cry:

**Comment by healther** [comment]
@alalazo If you meant me with the rebaseing, I will need some pointers on how to do that, I thought github automatically does that?

**Comment by adamjstewart** [comment]
@healther To rebase, you would do something like:
```
$ git checkout develop
$ git pull upstream develop  # assuming LLNL/spack is called upstream
$ git checkout feature/mosh
$ git rebase develop
# make sure everything looks good!
$ git push --force origin feature/mosh  # assuming electronicvisions/spack is called origin
```
We also have a section in the documentation on this subject:
http://spack.readthedocs.io/en/latest/contribution_guide.html#git-workflows

**Comment by citibeth** [comment]
Or use SmartGit.  There's a nice point-and-click way to rebase, at to check
your work visually.

**Comment by healther** [comment]
Let's see if this works out as planned

File: var/spack/repos/builtin/packages/mosh/package.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    keystrokes. Mosh is a replacement for SSH. It's more robust and responsive,
+    especially over Wi-Fi, cellular, and long-distance links.
+    """
+
+    homepage = "https://mosh.org/"
+    url      = "https://mosh.org/mosh-1.2.6.tar.gz"
+
+    version('1.2.6', 'bb4e24795bb135a754558176a981ee9e')
+
+    depends_on('perl')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mosh/package.py]: Is Perl a build dependency? I doubt it's a link dependency. May be a run dependency too if it's needed after installation.

**Comment by eschnett** [comment]
I have used `mosh` in the past with several HPC centres. This turned out to be a large problem: While mosh uses a regular ssh server for the initial connection and authentication, it then sets up its own daemon process to listen for incoming UDP traffic. This bypasses the standard Unix mechanisms that keep track of who is currently logged in, and from where the connection is made. At least one HPC centre mistook this for me attacking their site, and after clearing up this misunderstanding, still explicitly disallows using mosh for this reason.

I am now using a hybrid approach based on `autossh` (needs to be installed only locally on my laptop) and `tmux` (equivalent to GNU screen). (I'm remapping the escape character to ^G, since I want to be able to use ^A and ^B for cursor movement.)

**Comment by citibeth** [comment]
Yes, I was afraid that might be the case.  I would never run it on my home
computer, since my security policy is to keep all ports closed except SSH.

File: var/spack/repos/builtin/packages/mosh/package.py (hunk 2 above)
Comment (tgamblin) [inline var/spack/repos/builtin/packages/mosh/package.py]: Turns out it's only a run dependency, per the [build instructions](https://github.com/mobile-shell/mosh/wiki/Build-Instructions)

**Comment by tgamblin** [comment]
I switched the deptype for `perl` to `run`.  If the tests pass this LGTM.

**Pull request merged**